common/Resources/ti.internal/extensions/js/console.js

```diff
@@ -1,30 +1,160 @@
-	nativeInfo.call(console, formatWithOptions(kColorInspectOptions, ...args));
-};
+	_writeToConsole(level, string) {
+		if (this._groupIndent.length !== 0) {
+			if (string.includes('\n')) {
+				string = string.replace(/\n/g, `\n${this._groupIndent}`);
+			}
+			string = this._groupIndent + string;
+		}
 
-console.log = function (...args) {
-	nativeLog.call(console, formatWithOptions(kColorInspectOptions, ...args));
-};
+		// Support wrapping Ti.API (which retains log level)
+		if (this._apiModule) {
+			this._apiModule[level](string);
+		} else {
+			// Support Node.JS streams like stdout/stderr which don't have log levels
+			const stream = (level === 'warn' || level === 'error') ? this._stderr : this._stdout;
+			// TODO: Handle this._ignoreErrors by doing try/catch/finally, hanging error handlers
+			stream.write(string);
+		}
+	}
 
-console.warn = function (...args) {
-	nativeWarn.call(console, formatWithOptions(kNoColorInspectOptions, ...args));
-};
+	info(...args) {
+		this._writeToConsole('info', formatWithOptions(kColorInspectOptions, ...args));
+	}
+
+	warn(...args) {
+		this._writeToConsole('warn', formatWithOptions(kNoColorInspectOptions, ...args));
+	}
+
+	error(...args) {
+		this._writeToConsole('error', formatWithOptions(kNoColorInspectOptions, ...args));
+	}
+
+	debug(...args) {
+		this._writeToConsole('debug', formatWithOptions(kColorInspectOptions, ...args));
+	}
+
+	clear() {} // no-op
+
+	group(...data) {
+		if (data.length > 0) {
+			this.log(...data);
+		}
+		this._groupIndent += '  ';
+	}
+
+	groupEnd() {
+		this._groupIndent = this._groupIndent.slice(0, this._groupIndent.length - 2);
+	}
+
+	dir(obj, options) {
+		this._writeToConsole('info', inspect(obj, {
+			customInspect: false,
+			...options
+		}));
+	}
+
+	assert(value, ...args) {
+		if (!value) {
+			args[0] = `Assertion failed${args.length === 0 ? '' : `: ${args[0]}`}`;
+			this.warn(...args);  // The arguments will be formatted in warn() again
+		}
+	}
+
+	count(label = 'default') {
+		// Ensures that label is a string, and only things that can be
+		// coerced to strings. e.g. Symbol is not allowed
+		label = `${label}`;
+		let count = this._counts.get(label);
+		if (count === undefined) {
+			count = 1;
+		} else {
+			count++;
+		}
+		this._counts.set(label, count);
+		this.log(`${label}: ${count}`);
+	}
+
+	countReset(label = 'default') {
+		if (!this._counts.has(label)) {
+			process.emitWarning(`Count for '${label}' does not exist`);
+			return;
+		}
+		this._counts.delete(`${label}`);
+	}
+
+	time(label = 'default') {
+		label = `${label}`;
+		if (this._times.has(label)) {
+			process.emitWarning(`Label ${label}" already exists`);
+			return;
+		}
+		this._times.set(label, Date.now());
+	}
+
+	timeEnd(label = 'default') {
+		const warned = logTime(this, label);
+		if (!warned) {
+			this._times.delete(label);
+		}
+	}
+
+	timeLog(label = 'default', ...logData) {
+		logTime(this, label, logData);
+	}
+	// TODO: console.table()
```

Comment:
I was wondering if we should Implement `table()` as a no-op, printing a one-time warning that it is not implemented yet. Just to be API compatible and reduce possible crashes with third-party modules that actually try to use it.


Something like this:
```js
table() {
  if (!table.warned) {
    table.warned = true;
    process.emitWarning('"console.table" is not yet implemented in Titanium!');
  }
}
```